Fixes #18070 - add_record() checks for conflicts with leases now.

dmitri-d opened the pull request.

dmitri-d committed:
Fixes #18070 - add_record() checks for conflicts with leases now.

dmitri-d commented:
@timogoebel: updated add_record to check for conflicts with leases.

dmitri-d commented:
[test]

dmitri-d commented:
[test]

timogoebel commented on modules/dhcp_common/server.rb:
```diff
@@ -118,44 +118,42 @@ def inspect
+      unless similar_records.empty?
+        logger.warn "Request to create a conflicting DHCP record"
+        logger.debug "request: #{to_return.inspect}"
+        logger.debug "existing:   #{similar_records.inspect}"
```
Too many whitespaces between `:` and `#`

timogoebel commented on modules/dhcp_common/record.rb:
```diff
@@ -34,25 +33,19 @@ def [] opt
-      end
-      true
+    def ==(other)
+      ip == other.ip && mac == other.mac && subnet == other.subnet && deleteable? == other.deleteable? && options == other.options
```
Should we check if other is really a record before comparison? Same for `Proxy::DHCP::Lease`.

timogoebel commented on modules/dhcp_common/server.rb:
```diff
@@ -118,44 +118,42 @@ def inspect
 
-      validate_ip(options[:ip])
-      validate_mac(options[:mac])
-      raise(Proxy::DHCP::Error, "Must provide hostname") unless options[:hostname]
```
Is this safe to remove because we use "name" as a fallback? Should we add a check if name is set?

timogoebel reviewed (commented):
@witlessbird : Thanks. I've left some inline comments!

dmitri-d commented:
- restored name presence check in add_record method
- fixed formatting in https://github.com/theforeman/smart-proxy/pull/498#discussion-diff-97348267R143
- rewrote equality methods to remove duplication

timogoebel reviewed (approved):
Looks good and works great, thanks @witlessbird.
